src/dx/parse_token.rs

```diff
@@ -0,0 +1,151 @@
+use crate::core::PubNubError;
+use base64::engine::general_purpose;
+use base64::Engine;
+use ciborium::de::from_reader;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::ops::Deref;
+
+/// The [`parse_token`] function decodes an existing token and returns the struct containing permissions embedded in that token.
+/// The client can use this method for debugging to check the permissions to the resources
+pub fn parse_token(token: &str) -> Result<Token, PubNubError> {
+    let token_bytes = general_purpose::URL_SAFE_NO_PAD
+        .decode(token.as_bytes())
+        .map_err(|e| PubNubError::TokenDeserializationError(e.to_string()))?;
+
+    from_reader(token_bytes.deref())
+        .map_err(|e| PubNubError::TokenDeserializationError(e.to_string()))
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(untagged)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub enum Token {
+    V2(TokenV2),
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[allow(dead_code)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct TokenV2 {
+    #[serde(rename = "v")]
+    version: u8,
+    #[serde(rename = "t")]
+    timestamp: u32,
+    #[serde(rename = "ttl")]
+    ttl: u32,
+    #[serde(rename = "uuid")]
+    authorized_user_id: Option<String>,
+    #[serde(rename = "res")]
+    resources: TokenResources,
+    #[serde(rename = "pat")]
+    patterns: TokenResources,
+    meta: HashMap<String, MetaValue>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[allow(dead_code)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct TokenResources {
+    #[serde(rename = "chan")]
+    channels: HashMap<String, ResourcePermissions>,
+    #[serde(rename = "grp")]
+    groups: HashMap<String, ResourcePermissions>,
+    #[serde(rename = "uuid")]
+    users: HashMap<String, ResourcePermissions>,
+}
+
+impl From<u8> for ResourcePermissions {
+    fn from(int: u8) -> Self {
+        ResourcePermissions {
+            read: int & 1 != 0,
+            write: int & 2 != 0,
+            manage: int & 4 != 0,
+            delete: int & 8 != 0,
+            create: int & 16 != 0,
+            get: int & 32 != 0,
+            update: int & 64 != 0,
+            join: int & 128 != 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[allow(dead_code)]
+#[serde(from = "u8")]
+pub struct ResourcePermissions {
+    read: bool,
+    write: bool,
+    manage: bool,
+    delete: bool,
+    create: bool,
+    get: bool,
+    update: bool,
+    join: bool,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(untagged)]
+pub enum MetaValue {
+    String(String),
+    Integer(i64),
+    Float(f64),
+    Bool(bool),
+    /// It's always [`None`]
+    Null(Option<String>),
```

Comment:
@kleewho what will happen, if `null` will be in the payload? 
With `meta` in grant token I had to do this:

```rust
pub enum GrantTokenMetaValue {
    String(String),
    Integer(i64),
    Float(f64),
    Boolean(bool),
    Null,
}

#[cfg(feature = "serde")]
impl serde::Serialize for GrantTokenMetaValue {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            GrantTokenMetaValue::String(val) => serializer.serialize_str(val),
            GrantTokenMetaValue::Integer(val) => serializer.serialize_i64(*val),
            GrantTokenMetaValue::Float(val) => serializer.serialize_f64(*val),
            GrantTokenMetaValue::Boolean(val) => serializer.serialize_bool(*val),
            GrantTokenMetaValue::Null => serializer.serialize_unit(),
        }
    }
}
```

This way, it properly encodes `GrantTokenMetaValue::Null` into `null` (JSON `null`, not string `"null"`).

Reply:
This will happen `("other".into(), Null(None)),`. It's in the test